tests/email/test_tokens.py

```diff
@@ -41,177 +41,205 @@ def password_reset_token():
-    data_api_client.get_user.assert_called_once_with(123)
+        with freeze_time('2016-01-01T12:00:00.30Z'):
+            data = decode_token(token, 'Secret', 'PassSalt')
 
+        assert data == ({'email': 'test@example.com', 'user': 123}, datetime(2016, 1, 1, 12, 0, 0))
 
-def test_decode_password_reset_token_does_not_work_if_bad_token(email_app, data_api_client, password_reset_token):
-    token = generate_token(password_reset_token, "Key", 'PassSalt')[1:]
 
-    with email_app.app_context():
-        assert decode_password_reset_token(token, data_api_client) == {'error': 'token_invalid'}
+class TestDecodePasswordReset:
 
+    def test_decode_password_reset_token_ok_for_good_token(self, email_app, data_api_client, password_reset_token):
+        with email_app.app_context():
+            token = generate_token(password_reset_token, "Key", 'PassSalt')
+            assert decode_password_reset_token(token, data_api_client) == password_reset_token
+        data_api_client.get_user.assert_called_once_with(123)
 
-def test_decode_password_reset_token_does_not_work_if_token_expired(email_app, data_api_client, password_reset_token):
-    with freeze_time('2015-01-02 03:04:05'):
-        # Token was generated a year before current time
-        token = generate_token(password_reset_token, "Key", 'PassSalt')
+    def test_decode_password_reset_token_does_not_work_if_bad_token(
+            self, email_app, data_api_client, password_reset_token):
+        token = generate_token(password_reset_token, "Key", 'PassSalt')[1:]
 
-    with freeze_time('2016-01-02 03:04:05'):
         with email_app.app_context():
             assert decode_password_reset_token(token, data_api_client) == {'error': 'token_invalid'}
 
+    @pytest.mark.parametrize(
+        'decode_time, expected_result', [
+            ('2016-01-02 03:04:05', 'ok'),
+            ('2016-01-03 03:04:05', 'ok'),
+            ('2016-01-03 03:04:06', 'error'),
+        ]
+    )
+    def test_decode_password_reset_token_is_only_valid_within_a_day_of_token_creation(
+        self, decode_time, expected_result, email_app, data_api_client, password_reset_token
+    ):
+        with email_app.app_context():
+            with freeze_time('2016-01-02 03:04:05'):
+                token = generate_token(password_reset_token, "Key", 'PassSalt')
 
-def test_decode_password_reset_token_does_not_work_if_password_changed_later_than_token(
-    email_app, data_api_client, password_reset_token
-):
-    with freeze_time('2016-01-01T11:00:00.30Z'):
-        # Token was generated an hour earlier than password was changed
-        token = generate_token(password_reset_token, "Key", 'PassSalt')
-
-    with freeze_time('2016-01-01T13:00:00.30Z'):
-        # Token is two hours old; password was changed an hour ago
         with email_app.app_context():
-            assert decode_password_reset_token(token, data_api_client) == {'error': 'token_invalid'}
+            with freeze_time(decode_time):
```

Comment:
Does this need to be in a different `email_app.app_context`?

Reply:
Good question. I'll try it out.